src/backend.rs

```diff
@@ -0,0 +1,48 @@
+use crate::{
+    api::{Reply, Request},
+    error::Error,
+    platform::Platform,
+    service::ServiceResources,
+    types::{Context, CoreContext},
+};
+
+pub enum BackendId<C> {
+    Software,
+    Custom(C),
+}
+
+pub trait Backend<P: Platform> {
+    type Context;
+
+    fn request(
+        &mut self,
+        core_ctx: &mut CoreContext,
+        backend_ctx: &mut Self::Context,
+        request: &Request,
+        resources: &mut ServiceResources<P>,
+    ) -> Result<Reply, Error> {
+        let _ = (core_ctx, backend_ctx, request, resources);
+        Err(Error::RequestNotAvailable)
+    }
+}
+
+pub trait Dispatch<P: Platform> {
+    type BackendId: 'static;
+    type Context: Default;
+
+    fn request(
+        &mut self,
+        backend: &Self::BackendId,
+        ctx: &mut Context<Self::Context>,
+        request: &Request,
+        resources: &mut ServiceResources<P>,
+    ) -> Result<Reply, Error> {
+        let _ = (backend, ctx, request, resources);
+        Err(Error::RequestNotAvailable)
+    }
+}
+
+impl<P: Platform> Dispatch<P> for () {
+    type BackendId = ();
```

Comment:
I would prefer using a void type `enum Empty {}` here instead of using the unit type (name to be bikeshed).
Cf. https://github.com/robin-nitrokey/trussed/compare/backends...nickray:trussed:elaboration even if the Dispatch method changed now. This enforces that you can't even generate a BackendId::Custom.

Reply:
Done.

Reply:
Could probably be `!` once it’s stable.